File: typescript/src/models/gemini.ts
/** Google Gemini adapter (Gemini API `generateContent`). */

import { ConfigError, ModelError } from "../errors.js";
import { Message, newId, Usage, type ModelRequest, type ModelResponse, type ToolCall } from "../types.js";
import { env, HTTPClient, ownRaw, type HTTPOptions, type Model, type ModelEvent } from "./base.js";

export const GEMINI_BASE_URL = "https://generativelanguage.googleapis.com/v1beta";

type Part = Record<string, any>;

export interface GeminiModelOptions extends HTTPOptions {
  apiKey?: string;
  baseUrl?: string;
  headers?: Record<string, string>;
}

export class GeminiModel implements Model {
  readonly provider = "gemini";
  readonly model: string;
  readonly http: HTTPClient;

  constructor(model: string, options: GeminiModelOptions = {}) {
    this.model = model.replace(/^models\//, "");
    const key = options.apiKey ?? env("GEMINI_API_KEY") ?? env("GOOGLE_API_KEY");
    if (!key) throw new ConfigError("gemini: set GEMINI_API_KEY or pass apiKey");
    const baseUrl = options.baseUrl ?? env("GEMINI_BASE_URL") ?? GEMINI_BASE_URL;
    this.http = new HTTPClient(baseUrl, { "x-goog-api-key": key, "Content-Type": "application/json", ...(options.headers ?? {}) }, "gemini", options);
  }

  buildBody(req: ModelRequest): Record<string, unknown> {
    const s = req.settings ?? {};
    const body: Record<string, unknown> = { contents: this.contents(req.messages) };
    const system = [req.system, ...req.messages.filter((m) => m.role === "system").map((m) => m.content)].filter(Boolean);
    if (system.length) body.systemInstruction = { parts: [{ text: system.join("\n\n") }] };
    if (req.tools?.length) {
      body.tools = [{ functionDeclarations: req.tools.map((t) => ({ name: t.name, description: t.description, parametersJsonSchema: t.parameters })) }];
      if (s.toolChoice) {
        const mode = ({ auto: "AUTO", required: "ANY", none: "NONE" } as Record<string, string>)[s.toolChoice];
        const cfg: Record<string, unknown> = { mode: mode ?? "ANY" };
        if (!mode) cfg.allowedFunctionNames = [s.toolChoice];
        body.toolConfig = { functionCallingConfig: cfg };
      }
    }
    const gen: Record<string, unknown> = {};
    if (s.temperature !== undefined) gen.temperature = s.temperature;
    if (s.topP !== undefined) gen.topP = s.topP;
    if (s.maxOutputTokens !== undefined) gen.maxOutputTokens = s.maxOutputTokens;
    if (req.outputSchema) {
      gen.responseMimeType = "application/json";
      gen.responseJsonSchema = req.outputSchema.schema;
    }
    if (Object.keys(gen).length) body.generationConfig = gen;
    Object.assign(body, s.extra ?? {});
    return body;
  }

  private contents(messages: Message[]): { role: string; parts: Part[] }[] {
    const out: { role: string; parts: Part[] }[] = [];
    const push = (role: string, parts: Part[]) => {
      const last = out[out.length - 1];
      if (last && last.role === role) last.parts.push(...parts);
      else out.push({ role, parts: [...parts] });
    };
    for (const m of messages) {
      if (m.role === "system") continue;
      if (m.role === "user") push("user", [{ text: m.content || " " }]);
      else if (m.role === "tool") {
        let result: unknown = m.content;
        try {
          result = JSON.parse(m.content);
        } catch {
          /* keep string */
        }
        const fr: Part = { name: m.name ?? "tool", response: m.isError ? { error: result } : { result } };
        if (m.toolCallId && !m.toolCallId.startsWith("gcall_")) fr.id = m.toolCallId;
        push("user", [{ functionResponse: fr }]);
      } else {
        const raw = ownRaw(this.provider, m) as Part[] | undefined;
        if (raw) {
          push("model", raw.map((p) => ({ ...p })));
          continue;
        }
        const parts: Part[] = [];
        if (m.content) parts.push({ text: m.content });
        for (const tc of m.toolCalls ?? []) {
          const fc: Part = { name: tc.name, args: tc.arguments };
          if (!tc.id.startsWith("gcall_")) fc.id = tc.id;
          parts.push({ functionCall: fc });
        }
        push("model", parts.length ? parts : [{ text: " " }]);
      }
    }
    return out;
  }

  private usage(u: any): Usage {
    return new Usage({ inputTokens: u?.promptTokenCount ?? 0, outputTokens: (u?.candidatesTokenCount ?? 0) + (u?.thoughtsTokenCount ?? 0), requests: 1 });
  }

  private toMessage(parts: Part[]): Message {
    const text = parts.filter((p) => "text" in p && !p.thought).map((p) => p.text).join("");
    const calls: ToolCall[] = parts
      .filter((p) => p.functionCall)
      .map((p) => ({ id: p.functionCall.id ?? newId("gcall"), name: p.functionCall.name, arguments: p.functionCall.args ?? {} }));
    const msg = Message.assistant(text, calls);
    msg.raw = { provider: this.provider, data: parts };
    return msg;
  }

  private candidates(data: any): any[] {
    const c = data.candidates ?? [];
    if (!c.length && data.promptFeedback?.blockReason)
      throw new ModelError(`gemini: prompt blocked (${data.promptFeedback.blockReason})`, undefined, data, this.provider);
    return c;
  }

  async generate(req: ModelRequest): Promise<ModelResponse> {
    const data = await this.http.postJSON(`models/${this.model}:generateContent`, this.buildBody(req), req.signal);
    const cand = this.candidates(data)[0];
    return { message: this.toMessage(cand?.content?.parts ?? []), usage: this.usage(data.usageMetadata), stopReason: cand?.finishReason };
  }

  async *stream(req: ModelRequest): AsyncIterable<ModelEvent> {
    const parts: Part[] = [];
    let usage = new Usage({ requests: 1 });
    let finish: string | undefined;
    const plain = (p: Part | undefined) => p && "text" in p && !p.thought && !p.thoughtSignature;
    for await (const { data } of this.http.postSSE(`models/${this.model}:streamGenerateContent?alt=sse`, this.buildBody(req), req.signal)) {
      const chunk = JSON.parse(data);
      if (chunk.error) throw new ModelError(`gemini: ${JSON.stringify(chunk.error)}`, undefined, chunk, this.provider);
      if (chunk.usageMetadata) usage = this.usage(chunk.usageMetadata);
      const cand = this.candidates(chunk)[0];
      if (!cand) continue;
      finish = cand.finishReason ?? finish;
      for (const p of cand.content?.parts ?? []) {
        if (plain(p) && plain(parts[parts.length - 1])) parts[parts.length - 1].text += p.text;
        else parts.push({ ...p });
        if ("text" in p && !p.thought && p.text) yield { type: "text_delta", delta: p.text };
      }
    }
    const msg = this.toMessage(parts);
    for (const tc of msg.toolCalls ?? []) yield { type: "tool_call", ...tc };
    yield { type: "done", response: { message: msg, usage, stopReason: finish } };
  }
}

File: python/src/openharness/models/gemini.py
"""Google Gemini adapter (Gemini API ``generateContent``)."""

from __future__ import annotations

import json
import os
from collections.abc import AsyncIterator
from typing import Any

from ..errors import ConfigError, ModelError
from ..types import Message, ModelRequest, ModelResponse, ToolCall, Usage, new_id
from .base import HTTPClient, Model, ModelEvent

GEMINI_BASE_URL = "https://generativelanguage.googleapis.com/v1beta"


class GeminiModel(Model):
    provider = "gemini"

    def __init__(
        self,
        model: str,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        headers: dict[str, str] | None = None,
        timeout: float = 300.0,
        max_retries: int = 3,
    ):
        self.model = model.removeprefix("models/")
        key = api_key or os.environ.get("GEMINI_API_KEY") or os.environ.get("GOOGLE_API_KEY")
        if not key:
            raise ConfigError("gemini: set GEMINI_API_KEY or pass api_key=")
        self.base_url = base_url or os.environ.get("GEMINI_BASE_URL") or GEMINI_BASE_URL
        self.http = HTTPClient(
            self.base_url,
            {"x-goog-api-key": key, "Content-Type": "application/json", **(headers or {})},
            timeout=timeout,
            max_retries=max_retries,
            provider="gemini",
        )

    async def aclose(self) -> None:
        await self.http.aclose()

    # ------------------------------------------------------------- request

    def build_body(self, req: ModelRequest) -> dict[str, Any]:
        s = req.settings
        body: dict[str, Any] = {"contents": self._contents(req.messages)}
        system_parts = [req.system] if req.system else []
        system_parts += [m.content for m in req.messages if m.role == "system" and m.content]
        if system_parts:
            body["systemInstruction"] = {"parts": [{"text": "\n\n".join(system_parts)}]}
        if req.tools:
            body["tools"] = [
                {
                    "functionDeclarations": [
                        {"name": t.name, "description": t.description, "parametersJsonSchema": t.parameters}
                        for t in req.tools
                    ]
                }
            ]
            if s.tool_choice:
                mode = {"auto": "AUTO", "required": "ANY", "none": "NONE"}.get(s.tool_choice)
                cfg: dict[str, Any] = {"mode": mode or "ANY"}
                if mode is None:
                    cfg["allowedFunctionNames"] = [s.tool_choice]
                body["toolConfig"] = {"functionCallingConfig": cfg}
        gen: dict[str, Any] = {}
        if s.temperature is not None:
            gen["temperature"] = s.temperature
        if s.top_p is not None:
            gen["topP"] = s.top_p
        if s.max_output_tokens is not None:
            gen["maxOutputTokens"] = s.max_output_tokens
        if req.output_schema:
            gen["responseMimeType"] = "application/json"
            gen["responseJsonSchema"] = req.output_schema.schema
        if gen:
            body["generationConfig"] = gen
        body.update(s.extra)
        return body

    def _contents(self, messages: list[Message]) -> list[dict[str, Any]]:
        out: list[dict[str, Any]] = []

        def push(role: str, parts: list[dict[str, Any]]) -> None:
            if out and out[-1]["role"] == role:
                out[-1]["parts"].extend(parts)
            else:
                out.append({"role": role, "parts": list(parts)})

        for m in messages:
            if m.role == "system":
                continue
            if m.role == "user":
                push("user", [{"text": m.content or " "}])
            elif m.role == "tool":
                try:
                    result: Any = json.loads(m.content)
                except (ValueError, TypeError):
                    result = m.content
                response = {"error": result} if m.is_error else {"result": result}
                fr: dict[str, Any] = {"name": m.name or "tool", "response": response}
                if m.tool_call_id and not m.tool_call_id.startswith("gcall_"):
                    fr["id"] = m.tool_call_id
                push("user", [{"functionResponse": fr}])
            else:
                raw = self.own_raw(m)
                if raw:
                    push("model", [dict(p) for p in raw])
                    continue
                parts: list[dict[str, Any]] = []
                if m.content:
                    parts.append({"text": m.content})
                for tc in m.tool_calls:
                    fc: dict[str, Any] = {"name": tc.name, "args": tc.arguments}
                    if not tc.id.startswith("gcall_"):
                        fc["id"] = tc.id
                    parts.append({"functionCall": fc})
                push("model", parts or [{"text": " "}])
        return out

    # ------------------------------------------------------------- response

    @staticmethod
    def _usage(u: dict[str, Any] | None) -> Usage:
        u = u or {}
        out = (u.get("candidatesTokenCount") or 0) + (u.get("thoughtsTokenCount") or 0)
        return Usage(input_tokens=u.get("promptTokenCount") or 0, output_tokens=out, requests=1)

    def _to_message(self, parts: list[dict[str, Any]]) -> Message:
        text = "".join(p.get("text", "") for p in parts if "text" in p and not p.get("thought"))
        calls = []
        for p in parts:
            fc = p.get("functionCall")
            if fc:
                calls.append(
                    ToolCall(id=fc.get("id") or new_id("gcall"), name=fc["name"], arguments=fc.get("args") or {})
                )
        msg = Message.assistant(text, calls)
        msg.raw = {"provider": self.provider, "data": parts}
        return msg

    def _check(self, data: dict[str, Any]) -> list[dict[str, Any]]:
        cands = data.get("candidates") or []
        if not cands:
            fb = data.get("promptFeedback") or {}
            if fb.get("blockReason"):
                raise ModelError(f"gemini: prompt blocked ({fb['blockReason']})", body=data, provider=self.provider)
            return []
        return cands

    async def generate(self, request: ModelRequest) -> ModelResponse:
        data = await self.http.post_json(f"models/{self.model}:generateContent", self.build_body(request))
        cands = self._check(data)
        parts = (cands[0].get("content") or {}).get("parts") or [] if cands else []
        finish = cands[0].get("finishReason") if cands else None
        return ModelResponse(self._to_message(parts), self._usage(data.get("usageMetadata")), finish)

    async def stream(self, request: ModelRequest) -> AsyncIterator[ModelEvent]:
        parts: list[dict[str, Any]] = []
        usage = Usage(requests=1)
        finish = None
        async for _event, data in self.http.post_sse(
            f"models/{self.model}:streamGenerateContent?alt=sse", self.build_body(request)
        ):
            chunk = json.loads(data)
            if chunk.get("error"):
                raise ModelError(f"gemini: {chunk['error']}", body=chunk, provider=self.provider)
            if chunk.get("usageMetadata"):
                usage = self._usage(chunk["usageMetadata"])
            for cand in self._check(chunk)[:1]:
                finish = cand.get("finishReason") or finish
                for p in (cand.get("content") or {}).get("parts") or []:
                    if (
                        "text" in p
                        and not p.get("thought")
                        and not p.get("thoughtSignature")
                        and parts
                        and "text" in parts[-1]
                        and not parts[-1].get("thought")
                        and not parts[-1].get("thoughtSignature")
                    ):
                        parts[-1]["text"] += p["text"]  # merge plain text chunks
                    else:
                        parts.append(dict(p))
                    if "text" in p and not p.get("thought") and p["text"]:
                        yield ModelEvent("text_delta", {"delta": p["text"]})
        msg = self._to_message(parts)
        for tc in msg.tool_calls:
            yield ModelEvent("tool_call", tc.to_dict())
        yield ModelEvent("done", response=ModelResponse(msg, usage, finish))
